### ISSUE TITLE
Calling `setColumns` resets column widths, show/hide status, and column ordering

### ISSUE THREAD

**Issue by rathboma** [olifolkerd/tabulator#3959]
So this is a pretty straightforward one, but it's tricky to know what is the correct behavior.

I have a way for users to refresh the schema of a table (in case they made changes), but doing so causes us to call `setColumns`.

When our code calls that function, any memoized width or visibility data disappears, presumably so would column orderings.

Not sure what the best thing to do about this would be, maybe some sort of flag like `setColumns({ maintainWidths: true, maintainVisibility: true, ...})`?

![Peek 2022-09-28 15-20](https://user-images.githubusercontent.com/279769/192880792-5b7b0b23-ffb1-43a1-af54-8d3b6278b247.gif)

**Comment by olifolkerd** [comment]
Hey @rathboma 

That is the correct behaviour, `setColumns` completely removes the existing columns and replaces them, its whole purpose it to allow a complete reconfiguration of columns.

I would not be looking to add any props to this function, as it functions in the same way as the `setData` function, it is a signal to the table that that data has been completely replaced.

From your example, (based on what i can see in the gif) it would be entirely manageable by calling the `show` and `hide` functions on each of the column components without any need for calling set columns at all. If you use `blockRedraw` and `restoreRedraw` then it should have no affect on table performance.

If you want to do something more advanced then i could look at adding  a `replaceColumns` function, that works in a similar way to the `replaceData` function, where it adds in columns if the dont exist and updates ones that do. how does that sound to you.

We could make it so that it only updates props in existing columns that are set in the incomming data, so if you want to update one feature of a column but not another then just dont include it in the array of column definitions you pass to the function.

How does that sound?

Cheers

Oli :)

**Comment by rathboma** [comment]
Hi @olifolkerd,

Your idea for a `replaceColumns` feature sounds perfect. That would help me maintain quite a lot of state in the table views, like page, scroll position, etc.

How would that handle removed columns do you think? Would it remove them from the table if they don't exist in the `replaceColumns` payload?

**Comment by olifolkerd** [comment]
Yeah it would remove them if they didn't exist. If you wanted to keep one but not change anything you could just set the field prop on the object. How does that sound?

**Comment by rathboma** [comment]
Perfect
